core/tests/test_utils.py

```diff
@@ -1450,18 +1723,24 @@ class AppEngineTestBase(TestBase):
     """
 
     # Environment values that our tests depend on.
-    AUTH_DOMAIN = 'example.com'
-    HTTP_HOST = 'localhost'
-    SERVER_NAME = 'localhost'
-    SERVER_PORT = '8080'
-    DEFAULT_VERSION_HOSTNAME = '%s:%s' % (HTTP_HOST, SERVER_PORT)
+    AUTH_DOMAIN: Final = 'example.com'
+    HTTP_HOST: Final = 'localhost'
+    SERVER_NAME: Final = 'localhost'
+    SERVER_PORT: Final = '8080'
+    DEFAULT_VERSION_HOSTNAME: Final = '%s:%s' % (HTTP_HOST, SERVER_PORT)
 
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
         # Defined outside of setUp() because we access it from methods, but can
         # only install it during the run() method. Defining it in __init__
         # satisfies pylint's attribute-defined-outside-init warning.
-        self._platform_taskqueue_services_stub = TaskqueueServicesStub(self)
+        # TaskqueueServicesStub can only accept 'GenericTestBase' class but here
+        # we are providing super class (AppEngineTestBase) which causes MyPy to
+        # throw `incompatible argument type` error. Thus to avoid the error, we
+        # used cast here.
+        self._platform_taskqueue_services_stub = TaskqueueServicesStub(
+            cast(GenericTestBase, self)
+        )
```

Comment:
No, we cannot do this because in `TaskqueueServicesStub` we are defining `self._test_base` https://github.com/oppia/oppia/blob/cc8f6d73d492d77aada65e73d7f94d258746185d/core/tests/test_utils.py#L817
and the type of `self._test_base` needs to be `GenericTestBase` because in method `_task_handler` we using `get_new_csrf_token` and `post_task` method on `self._test_base` and these methods are defined in `GenericTestBase`.

Reply:
But then there is a mistake in the code, right? As we are passing `AppEngineTestBase` which should not work anyway?

Reply:
Did some changes PTAL!

Reply:
https://github.com/oppia/oppia/blob/5e1fde53bce397f00e7f047c7fca18e2e1088a85/core/jobs/job_test_utils.py#L49

In the above line we inheriting AppEngineTestBase:
and when we call Init of this class
https://github.com/oppia/oppia/blob/5e1fde53bce397f00e7f047c7fca18e2e1088a85/core/jobs/job_test_utils.py#L63


Then below self was provided as a AppEngineTestBase not generictestbase ( which prevent us from using assert )
https://github.com/oppia/oppia/blob/5e1fde53bce397f00e7f047c7fca18e2e1088a85/core/tests/test_utils.py#L1464
So that's why we have to use cast here.

Reply:
> But then there is a mistake in the code, right

maybe yes not sure 100 percent

Reply:
Hmm, can you try moving the initialization to `GenericTestBase` then?

Reply:
Done!, PTAL!

Reply:
After doing these changes I'm failing the backend tests, can please take a look at once?

Reply:
Hmm, interesting I think I remember this issue, please add a TODO and an issue for me to investigate this.

Reply:
TODO is filed ( #15922  ). Also, TODO is added inside the comment.